File: frontend/src/types/scan.ts
export enum ScanStatus {
  PENDING = "pending",
  RUNNING = "running",
  COMPLETED = "completed",
  FAILED = "failed",
  CANCELLED = "cancelled",
}

export enum PhaseStatus {
  PENDING = "pending",
  RUNNING = "running",
  COMPLETED = "completed",
  FAILED = "failed",
  SKIPPED = "skipped",
}

export interface ScanPhase {
  id: number;
  scan_id: number;
  phase_number: number;
  phase_name: string;
  status: PhaseStatus;
  tool_statuses: Record<string, string>;
  started_at: string | null;
  completed_at: string | null;
  duration_seconds: number | null;
  error_message: string | null;
  log_output: string | null;
}

export interface ScanConfig {
  harvester_sources?: string;
  amass_timeout_minutes?: number;
  masscan_rate?: number;
  masscan_ports?: string;
  nmap_scripts?: string;
  nuclei_severity?: string[];
  nuclei_rate_limit?: number;
  enable_zap?: boolean;
  enable_openvas?: boolean;
  push_to_defectdojo?: boolean;
}

export interface Scan {
  id: number;
  scan_uid: string;
  target_domain: string;
  status: ScanStatus;
  current_phase: number;
  config: ScanConfig;
  results_dir: string;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_by: number;
  created_at: string;
  updated_at: string;
  phases: ScanPhase[];
}

export interface ScanCreate {
  target_domain: string;
  config?: ScanConfig;
}

File: backend/app/schemas/scan.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.base import PhaseStatus, ScanStatus


class ScanConfig(BaseModel):
    harvester_sources: str = "bing,crtsh,dnsdumpster"
    amass_timeout_minutes: int = Field(default=15, ge=1, le=60)
    masscan_rate: int = Field(default=10000, ge=100, le=100000)
    masscan_ports: str = "1-65535"
    nmap_scripts: str = "default,vuln"
    nuclei_severity: list[str] = Field(default=["low", "medium", "high", "critical"])
    nuclei_rate_limit: int = Field(default=150, ge=10, le=1000)
    enable_zap: bool = True
    enable_openvas: bool = False
    push_to_defectdojo: bool = True


class ScanCreate(BaseModel):
    target_domain: str = Field(..., min_length=3, max_length=255)
    config: ScanConfig = Field(default_factory=ScanConfig)


class ScanPhaseResponse(BaseModel):
    id: int
    phase_number: int
    phase_name: str
    status: PhaseStatus
    tool_statuses: dict = {}
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_seconds: float | None = None
    error_message: str | None = None

    model_config = {"from_attributes": True}


class ScanResponse(BaseModel):
    id: int
    scan_uid: str
    target_domain: str
    status: ScanStatus
    current_phase: int
    config: dict
    results_dir: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    error_message: str | None = None
    created_by: int
    created_at: datetime
    updated_at: datetime
    phases: list[ScanPhaseResponse] = []

    model_config = {"from_attributes": True}


class ScanListResponse(BaseModel):
    items: list[ScanResponse]
    total: int
    page: int
    per_page: int


class ScanLogResponse(BaseModel):
    scan_id: int
    logs: list[dict]
